manticore/platforms/evm_world_state.py

```diff
@@ -0,0 +1,468 @@
+import logging
+from abc import ABC, abstractmethod
+from eth_typing import ChecksumAddress, URI
+from io import TextIOBase
+from typing import Dict, List, Optional, Set, Tuple, Union
+from urllib.parse import ParseResult, urlparse
+from web3 import Web3
+from ..core.smtlib import Array, ArrayVariable, BitVec, BitVecConstant, BitVecITE, BitVecZeroExtend, ConstraintSet
+from ..ethereum.state import State
+from ..exceptions import EthereumError
+
+logger = logging.getLogger(__name__)
+
+
+# sam.moelius: map records which (symbolic) offsets have been written to.  data holds the values
+# written.
+
+
+class Storage:
+    def __init__(self, constraints: ConstraintSet, address: int):
+        """
+        :param constraints: the ConstraintSet with which this Storage object is associated
+        :param address: the address that owns this storage
+        """
+        self._data = constraints.new_array(
+            index_bits=256,
+            value_bits=256,
+            name=f"STORAGE_DATA_{address:x}",
+            avoid_collisions=True,
+            # sam.moelius: The use of default here creates unnecessary if-then-elses.  See
+            # ArrayProxy.get in expression.py.
+            # default=0,
+        )
+
+    def __getitem__(self, offset: Union[int, BitVec]) -> Union[int, BitVec]:
+        return self.get(offset, 0)
+
+    def get(self, offset: Union[int, BitVec], default: Union[int, BitVec]) -> Union[int, BitVec]:
+        if not isinstance(default, BitVec):
+            default = BitVecConstant(256, default)
+        return self._data.get(offset, default)
+
+    def set(self, offset: Union[int, BitVec], value: Union[int, BitVec]):
+        self._data[offset] = value
+
+    def get_items(self) -> List[Tuple[Union[int, BitVec], Union[int, BitVec]]]:
+        items = []
+        array = self._data.array
+        while not isinstance(array, ArrayVariable):
+            items.append((array.index, array.value))
+            array = array.array
+        return items
+
+    def dump(self, stream: TextIOBase, state: State):
+        concrete_indexes = set()
+        for sindex in self._data.written:
+            concrete_indexes.add(state.solve_one(sindex, constrain=True))
+
+        for index in concrete_indexes:
+            stream.write(
+                f"storage[{index:x}] = {state.solve_one(self._data[index], constrain=True):x}\n"
+            )
+
+
+####################################################################################################
+
+
+class WorldState:
+    def new_storage(
```

Comment:
What would be the docstring of this method? Is this a convenience method?

I'm struggling to think when do we want a storage without an account.
I expect a world state to have add_account(), delete_account(), accounts()

Reply:
> What would be the docstring of this method? Is this a convenience method?

I added the docstring. It is a convenience method.
https://github.com/trailofbits/manticore/blob/235d0b5796324956b332cceb69778f88d74688f2/manticore/platforms/evm_world_state.py#L80-L85

> I'm struggling to think when do we want a storage without an account.
> I expect a world state to have add_account(), delete_account(), accounts()

I don't think those methods make sense for the remote model, because they presume that an account "exists" or "does not exist". What is more accurate is there are things that we know about an address and things that we do not, which could include nonce, balance, storage, and code. That is why I espouse the removal of the `accounts` method.

Reply:
They do not make sense for the special case of a remote world. I guess those will not be implemented there.  And the layer on top of it will _need_ to handle that functionality. 

World State looks like something that needs a  way to add, delete and list account (or to check if it contains an account)

Reply:
> And the layer on top of it will _need_ to handle that functionality.

Why do you think those methods are needed?

Reply:
A method to add an account? So the user can add accounts to the ethereum world.
A method to del an account? I guess selfdestruct wants it.
A method to list accounts?  To concretize target addresses for example.
A method to check if an account exists? You need to know if the account is already there to calculate gas in tx instructions sometimes.